_posts/2021-02-09-an-introduction-to-soliditys-fuzz-testing-approach.md

```diff
@@ -0,0 +1,164 @@
+---
+layout: post
+published: true
+title: 'An Introduction to Solidity's Fuzz Testing Approach'
+date: '2021-02-09'
+author: Bhargava Shastry
+category: Explainers
+---
+
+Security vulnerabilities and bugs detract from software quality.
+To discover them early, we have adopted fuzz testing: feeding randomly generated programs to the Solidity compiler and observing the compilation runtime and code generated.
+
+Since Q1 2019, the Solidity compiler is fuzz tested via Google's [open-source software fuzz][1] (oss-fuzz) framework.
+
+In this post, we briefly describe the work that has been done on this front, and work that is currently in progress.
+
+### Fuzzer Overview
+
+Broadly speaking, we have developed two classes of fuzzers:
+  - Front-end fuzzers that test the compiler front-end (parser/analyser).
+  - Back-end fuzzers that test the compiler back-end (optimiser).
+
+### Front-end Fuzzers
+
+We have developed front-end fuzzers in order to test if programs are parsed and analysed correctly.
+Front-end fuzzing comprises mutating existing test cases (most often unit tests) in order to test how parsing and program analysis respond to corner case input.
+
+#### Example bug found by front-end fuzzer
+
+Solidity implements the [NatSpec format][3] for code documentation.
+The front-end fuzzer generated the following program that led to an assertion failure in the compiler (see Notes for more information):
+
+```
+contract C {
+    ///@return
+    modifier m22 {}
+}
+
+contract D is C {
+    modifier m22 {}
+}
+```
+
+The program was presumably mutated from the following perfectly valid test program.
+
+```
+contract C {
+    modifier m22 { _; }
+}
+contract D is C {
+}
+```
+
+The mutations prepended the `///@return` line to the modifier declaration and copied the modifier declaration to the derived contract.
+
+A natural question is the following: How did the fuzzer "know" that this particular mutation will lead to an error?
+The answer is that it does not know; the fuzzer simply mutates existing input to create new inputs and it just so happened that this mutation discovered a bug.
+A follow-up question could be: How are mutations created?
+Broadly speaking, mutations fall into one of the following categories:
+  - Flipping bit(s) at certain position(s) in the input stream.
```

Comment:
Does the singular make sense for these? It at least should be `a bit`, so maybe just stick with `bits`?

Reply:
Done